### PR TITLE
Document the BooleanBiFunction interface

### PR DESCRIPTION
**Pull request by haykam821**
Relates to #2389

### CODE REVIEW

**Review by liach — CHANGES_REQUESTED** [review]
Imo we should only document it as a part of the voxel shape system, as this (actually a boolean binary operator) is exclusively used in the voxel shape combination logic there. I would put this on hold.

File: mappings/net/minecraft/util/function/BooleanBiFunction.mapping [hunk 1]
```diff
@@ -1,17 +1,34 @@
...
 	FIELD field_16895 TRUE Lnet/minecraft/class_247;
+		COMMENT A {@link BooleanBiFunction} that always returns {@code true}.
 	FIELD field_16896 AND Lnet/minecraft/class_247;
+		COMMENT A {@link BooleanBiFunction} that returns {@code true} if both arguments are {@code true}.
 	FIELD field_16897 FALSE Lnet/minecraft/class_247;
+		COMMENT A {@link BooleanBiFunction} that always returns {@code false}.
 	FIELD field_16898 NOT_SECOND Lnet/minecraft/class_247;
+		COMMENT A {@link BooleanBiFunction} that returns {@code true} if the second argument is {@code false}.
 	FIELD field_16899 CAUSED_BY Lnet/minecraft/class_247;
+		COMMENT A {@link BooleanBiFunction} that returns {@code true} if the first argument is {@code true} and the second argument is {@code false}.
```
Comment (liach) [inline mappings/net/minecraft/util/function/BooleanBiFunction.mapping]:
```suggestion
		COMMENT A {@link BooleanBiFunction} that returns {@code true} if the first argument is {@code true} or the second argument is {@code false}.
```

File: mappings/net/minecraft/util/function/BooleanBiFunction.mapping [hunk 2]
```diff
@@ -1,17 +1,34 @@
...
 	FIELD field_16885 NOT_OR Lnet/minecraft/class_247;
+		COMMENT A {@link BooleanBiFunction} that returns {@code true} if neither argument is {@code true}.
 	FIELD field_16886 ONLY_FIRST Lnet/minecraft/class_247;
+		COMMENT A {@link BooleanBiFunction} that returns {@code true} if only the first argument is {@code true}.
 	FIELD field_16887 SECOND Lnet/minecraft/class_247;
+		COMMENT A {@link BooleanBiFunction} that returns {@code true} if the second argument is {@code true}.
 	FIELD field_16888 NOT_AND Lnet/minecraft/class_247;
+		COMMENT A {@link BooleanBiFunction} that returns {@code true} if either argument is {@code false}.
 	FIELD field_16889 CAUSES Lnet/minecraft/class_247;
+		COMMENT A {@link BooleanBiFunction} that returns {@code true} if the first argument is {@code false} and the second argument is {@code true}.
```
Comment (liach) [inline mappings/net/minecraft/util/function/BooleanBiFunction.mapping]:
```suggestion
		COMMENT A {@link BooleanBiFunction} that returns {@code true} if the first argument is {@code false} or the second argument is {@code true}.
```

File: mappings/net/minecraft/util/function/BooleanBiFunction.mapping [hunk 3]
```diff
@@ -1,17 +1,34 @@
 CLASS net/minecraft/class_247 net/minecraft/util/function/BooleanBiFunction
+	COMMENT A function that is provided two booleans and returns one boolean.
 	FIELD field_1366 OR Lnet/minecraft/class_247;
+		COMMENT A {@link BooleanBiFunction} that returns {@code true} if either argument is {@code true}.
 	FIELD field_16885 NOT_OR Lnet/minecraft/class_247;
+		COMMENT A {@link BooleanBiFunction} that returns {@code true} if neither argument is {@code true}.
 	FIELD field_16886 ONLY_FIRST Lnet/minecraft/class_247;
+		COMMENT A {@link BooleanBiFunction} that returns {@code true} if only the first argument is {@code true}.
 	FIELD field_16887 SECOND Lnet/minecraft/class_247;
+		COMMENT A {@link BooleanBiFunction} that returns {@code true} if the second argument is {@code true}.
```
Comment (liach) [inline mappings/net/minecraft/util/function/BooleanBiFunction.mapping:10]:
```suggestion
		COMMENT A {@link BooleanBiFunction} that returns the second argument.
```

File: mappings/net/minecraft/util/function/BooleanBiFunction.mapping [hunk 4]
```diff
@@ -1,17 +1,34 @@
...
 	FIELD field_16888 NOT_AND Lnet/minecraft/class_247;
+		COMMENT A {@link BooleanBiFunction} that returns {@code true} if either argument is {@code false}.
 	FIELD field_16889 CAUSES Lnet/minecraft/class_247;
+		COMMENT A {@link BooleanBiFunction} that returns {@code true} if the first argument is {@code false} and the second argument is {@code true}.
 	FIELD field_16890 SAME Lnet/minecraft/class_247;
+		COMMENT A {@link BooleanBiFunction} that returns {@code true} if both arguments are {@code true} or both arguments are {@code false}.
 	FIELD field_16891 FIRST Lnet/minecraft/class_247;
+		COMMENT A {@link BooleanBiFunction} that returns {@code true} if the first argument is {@code true}.
```
Comment (liach) [inline mappings/net/minecraft/util/function/BooleanBiFunction.mapping:18]:
```suggestion
		COMMENT A {@link BooleanBiFunction} that returns the first argument.
```

**Review by liach — CHANGES_REQUESTED** [review]
After looking, I think we should have this for now; though this is mainly used for combination in bit set voxel set, having something is better.

The main problems right now is you messed up the cause logic.

Additional suggestions: name the parameters in `apply` `a` `b` and use `a` `b` than "first argument" "second argument"; simplify the wording for `FIRST` and `SECOND`

File: mappings/net/minecraft/util/function/BooleanBiFunction.mapping (hunk 4 above)
Comment (haykam821) [reply mappings/net/minecraft/util/function/BooleanBiFunction.mapping:18]: I think that it is more clear to state the output in terms of the input without explicitly rewriting the expression in sentence form. This may be applicable to other, more complex data types, but not booleans.

File: mappings/net/minecraft/util/function/BooleanBiFunction.mapping (hunk 3 above)
Comment (haykam821) [reply mappings/net/minecraft/util/function/BooleanBiFunction.mapping:10]: See [here](https://github.com/FabricMC/yarn/pull/2493#discussion_r654747421)

File: mappings/net/minecraft/util/function/BooleanBiFunction.mapping (hunk 4 above)
Comment (liach) [reply mappings/net/minecraft/util/function/BooleanBiFunction.mapping:18]: If you want accuracy, should we use `returns ... if and only if` in place of all `returns ... if`?

Comment (haykam821) [reply mappings/net/minecraft/util/function/BooleanBiFunction.mapping:18]: It could be `returns ... exclusively if` as well. I assumed all other cases returning `false` would be implied but I can clarify it if necessary.

Comment (liach) [reply mappings/net/minecraft/util/function/BooleanBiFunction.mapping:18]: Then the current form should be good. Just make sure you fix the "causes" and "caused by" cases. a -> b is a causes b, or `!a || b` (like `a ? b : true`).

File: mappings/net/minecraft/util/function/BooleanBiFunction.mapping (hunk 1 above)
Comment (liach) [reply mappings/net/minecraft/util/function/BooleanBiFunction.mapping]:
```suggestion
		COMMENT A {@link BooleanBiFunction} that returns {@code true} if the first argument is {@code true} or the second argument is {@code false}. Equivalent to {@code second ? first : true}.
```

File: mappings/net/minecraft/util/function/BooleanBiFunction.mapping (hunk 2 above)
Comment (liach) [reply mappings/net/minecraft/util/function/BooleanBiFunction.mapping]:
```suggestion
		COMMENT A {@link BooleanBiFunction} that returns {@code true} if the first argument is {@code false} or the second argument is {@code true}. Equivalent to {@code first ? second : true}.
```

**Review by liach — CHANGES_REQUESTED** [review]

**Comment by liach** [comment]
Also, can you consider naming the parameters of the `apply(boolean, boolean)` method?

**Comment by liach** [comment]
Still, I am just requesting fixing `and` with `or` for `CAUSED` and `CAUSED_BY`. In those ones, if the premise is false, then the statement is vacuously true. @haykam821 mind check that out? See https://en.wikipedia.org/wiki/Vacuous_truth#Scope_of_the_concept (that first sentence is a good summary)

**Comment by liach** [comment]
While the rest of the changes are always good, you should fix the logical errors in the `CAUSED` and `CAUSED_BY` docs.

**Comment by liach** [comment]
@haykam821 Mind apply https://github.com/FabricMC/yarn/pull/2493#discussion_r654740944 and https://github.com/FabricMC/yarn/pull/2493#discussion_r654740925 suggestions? These are logical errors in your documentation. Otherwise, this is good to go.

**Comment by haykam821** [comment]
I used the following script to verify these inputs:

```js
(() => {
    const inputs = [
        [false, false],
        [false, true],
        [true, false],
        [true, true]
    ];
    const biFunction = ((a, b) => !a || b);

    return inputs.reduce((acc, [a, b]) => {
        const label = (a + "")[0] + (b + "")[0];
        acc[label] = biFunction(a, b);
        return acc;
    }, {});
})();
```

**Review by liach — APPROVED** [review]
